JEP389 Foreign Linker API: DownCall (Phase 1 / Primitive support)

ChengJin01 opened the pull request:
The prototype enables the downcall handle to the C function
for primitive types on X86_64 and Aarch64 by implementing a
lightweight ffi_call based solution in OpenJ9 to substitute
the counterpart APIs in OpenJDK.

Signed-off-by: Cheng Jin <jincheng@ca.ibm.com>

ChengJin01 commented:
### The high-level design of downcall
As compared to the conventional JNI native call, the downcall prototype aims to implement a lightweight `ffi_call` based solution to achieve the same behavior.

To be specific, `ProgrammableInvoker` in OpenJDK is entirely replaced with our own implementation, which wraps up a method handle to the underlying C code via `ffi_call`. In order to achieve this, the argument & return layouts at C level are mapped to the `ffi_type`  and the requested prep_cif is generated ahead of time via the out-of-line code and stored back to Java (with the support from the preset field’s offset in the static initialization of `ProgrammableInvoker`) prior to the invocation of method handle to ffi_call. 

To void duplicate memory allocation on the same argument layouts & unnecessary native invocation to the out-of-line code for the same argument & return layouts in the case of multiple calls and multithreading, hashmaps are created at Java level to cache the unique `prep_cif` and argument layouts, in which case:
1)	If `prep_cif` exists for the same arguments/return layouts,  then there is no need to call the out-of-line code;
2)	If not the case, check whether the same argument layouts exist:
1)	If so, only create `prep_cif` for the function with the existing argument `ffi_type` previously created for the same argument layouts (no need to allocate memory for the `ffi_type` for the same argument layouts);
2)	Otherwise, allocate memory for the new argument ff_types and create `prep_cif` afterwards

Different from RI which has no support on boolean, our prototype provides support on all primitive types (boolean/char/byte/short/int/long/float/double) as well as the corresponding primitive class types including Boolean, Character, Byte, Short, Integer, Long, Float and Double by converting the arguments/return value into base types internally in our code.

### Main changes involved in code (excluding other files with minor changes)
1) jdk/internal/foreign/abi/ProgrammableInvoker.java
The implementation of `ProgrammableInvoker` in OpenJ9 specific to downcall
2)/runtime/vm/BytecodeInterpreter.hpp
`inlProgrammableInvokerInvokeNative`  (the invocation of the method handle to ffi_call)
3) /runtime/vm/OutOfLineINL_jdk_internal_foreign_abi_ProgrammableInvoker.cpp
`OutOfLineINL_jdk_internal_foreign_abi_ProgrammableInvoker_resolveRequiredFields` (set up the offset required fields of ProgrammableInvoker in the static initialization)
`OutOfLineINL_jdk_internal_foreign_abi_ProgrammableInvoker_initCifNativeThunkData` (generate prep_cif with the existing/allocated cif_types)
4) /runtime/vm/FFITypeHelpers.hpp
`MapPrimitiveLayoutToFFIType` (convert the layout string at C level to the ffi_type via getPrimitiveFFIType)
`getPrimitiveFFIType` (obtain the ffi_type from the layout’s kind)

### References:
[1] https://openjdk.java.net/jeps/389
[2]https://docs.oracle.com/en/java/javase/16/docs/api/jdk.incubator.foreign/jdk/incubator/foreign/package-summary.html
[3] https://github.com/openjdk/panama-foreign/blob/d0e35dea8f77de9729972ba6ebd84d46f3807e3d/doc/panama_ffi.md
[4] https://github.com/openjdk/panama-foreign/blob/d0e35dea8f77de9729972ba6ebd84d46f3807e3d/doc/panama_memaccess.md

ChengJin01 commented:
Verified with newly added test cases in personal builds/locally compiled builds and jtreg test cases related to the primitive types at https://github.com/eclipse/openj9/issues/11195

gacholio commented on runtime/oti/VMHelpers.hpp:
```diff
@@ -186,6 +186,9 @@ typedef enum {
 	J9_BCLOOP_SEND_TARGET_METHODHANDLE_LINKTOINTERFACE,
 #endif /* defined(J9VM_OPT_OPENJDK_METHODHANDLE) */
+#if JAVA_SPEC_VERSION >= 16
+J9_BCLOOP_SEND_TARGET_INL_PROGRAMMABLEINVOKER_INVOKENATIVE,
```
Indent please.

ChengJin01 replied:
Fixed.

tajila commented on jcl/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableInvoker.java:
```diff
@@ -0,0 +1,461 @@
+	final static Lookup lookup = MethodHandles.lookup();
+	
+	/* The prep_cif and the corresponding argument layouts are cached & shared in multiple downcalls/threads */
+	private static final ConcurrentHashMap<String, Long> cachedCifNativeThunkAddr = new ConcurrentHashMap<String, Long>();
```
do you still need these to be concurrentHashMAps if you are using the `privateClassLock`?

ChengJin01 replied:
Updated with `HashMap` and verified with the test cases.

tajila commented on runtime/vm/OutOfLineINL_jdk_internal_foreign_abi_ProgrammableInvoker.cpp:
```diff
@@ -0,0 +1,162 @@
+			goto done;
+		}
+
+		memset(argTypes, 0, sizeof(ffi_type *) * argTypesCount);
```
Since you are writing all the elements in `argTypes` I dont think you need the memset

ChengJin01 replied:
Agreed and removed `memset()` as suggested.

tajila commented on runtime/vm/OutOfLineINL_jdk_internal_foreign_abi_ProgrammableInvoker.cpp:
```diff
@@ -0,0 +1,162 @@
+#if JAVA_SPEC_VERSION >= 16
+/* jdk.internal.foreign.abi.ProgrammableInvoker: private static synchronized native void resolveRequiredFields(); */
+VM_BytecodeAction
+OutOfLineINL_jdk_internal_foreign_abi_ProgrammableInvoker_resolveRequiredFields(J9VMThread *currentThread, J9Method *method)
```
I dont fully understand why this is needed, I was under the impression that this would be done in `initializeKnownClasses`

ChengJin01 replied:
Nope, this can't be accomplished by `initializeKnownClasses()` in the initailization stage because the `jclconstantpool` is initialized too early and can only depend on `java.base` which `ProgrammableInvoker` doesn't belong to, as discussed with @pshipton.

Another alternative is move `ProgrammableInvoker` to `java.base` with a different package & name. But it will end up with tons of unnecessary changes on OpenJDK side in replacing `ProgrammableInvoker` with the new package/name given `ProgrammableInvoker`  is used everywhere on all platforms.  To avoid this case, we chose to resolve these offsets in the static initializer.

tajila commented on runtime/vm/OutOfLineINL_jdk_internal_foreign_abi_ProgrammableInvoker.cpp:
```diff
@@ -0,0 +1,162 @@
+
+		cifArgTypesNode->argumentTypes = (void **)argTypes;
+		omrthread_monitor_enter(vm->cifArgumentTypesMutex);
+		J9_LINKED_LIST_ADD_LAST(vm->cifArgumentTypesListHead, cifArgTypesNode);
```
why is `cifArgumentTypesListHead` a linked list? it doesn't seem like we ever iterate it. Can it be a pool instead? Something with less memory overhead (e.g J9ThreadMonitorPool)

ChengJin01 replied:
Yes we never use the inserted node in the list anywhere in code as the only reason for this is to release all the memory in case of memory leak when VM exits unexpectedly. If pool works in the same way with less overhead, we should change to that.

ChengJin01 replied:
Replaced with `J9Pool *cifArgumentTypesCache` as suggested.

tajila commented on runtime/oti/VMHelpers.hpp:
```diff
@@ -186,6 +186,9 @@ typedef enum {
 	J9_BCLOOP_SEND_TARGET_METHODHANDLE_LINKTOINTERFACE,
 #endif /* defined(J9VM_OPT_OPENJDK_METHODHANDLE) */
+#if JAVA_SPEC_VERSION >= 16
+J9_BCLOOP_SEND_TARGET_INL_PROGRAMMABLEINVOKER_INVOKENATIVE,
```
ignore this one

ChengJin01 replied:
Already fixed previously.

tajila commented on runtime/vm/OutOfLineINL_jdk_internal_foreign_abi_ProgrammableInvoker.cpp:
```diff
@@ -0,0 +1,162 @@
+
+/* jdk.internal.foreign.abi.ProgrammableInvoker: private native void initCifNativeThunkData(String[] argLayouts, String retLayout, boolean newArgTypes); */
+VM_BytecodeAction
+OutOfLineINL_jdk_internal_foreign_abi_ProgrammableInvoker_initCifNativeThunkData(J9VMThread *currentThread, J9Method *method)
```
can you add a comment header briefly describing whats happening here. Also, describe the java args.

ChengJin01 replied:
Will update accordingly soon.

ChengJin01 replied:
Updated the comment as suggested above.

ChengJin01 commented on jcl/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableInvoker.java:
```diff
@@ -0,0 +1,461 @@
+			/* Replace the original handle with the specified types of the C function */
+			boundHandle = permuteMH(boundHandle, functionMethodType);
+			return boundHandle;
+		} catch (NoSuchMethodException | IllegalAccessException | NullPointerException | WrongMethodTypeException e) {
```
I will try to compile with only `ReflectiveOperationException` to see how it goes and will update if it works this way.

ChengJin01 replied:
Updated with `ReflectiveOperationException` and verified with the test cases.

ChengJin01 commented on jcl/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableInvoker.java:
```diff
@@ -0,0 +1,461 @@
+	
+	/* Collect and convert the passed-in arguments to an Object array for the underlying native call */
+	private static MethodHandle permuteMH(MethodHandle targetHandle, MethodType nativeMethodType) throws NullPointerException, WrongMethodTypeException {
+		MethodHandle resultHandle = targetHandle.asCollector(0, Object[].class, nativeMethodType.parameterCount());
```
It should work with `asCollector​(Class<?> arrayType, int arrayLength)` . I will double-check to confirm that.

ChengJin01 replied:
Updated with `asCollector​(Class<?> arrayType, int arrayLength)` and verified with the test cases.

ChengJin01 commented on jcl/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableInvoker.java:
```diff
@@ -0,0 +1,461 @@
+	 */
+	@SuppressWarnings("boxing")
+	private static long convertArgToLongValue(Class<?> argTypeClass, Object argValue) {
+		Class<?> realArgTypeClass = unboxingPrimitiveClassType(argTypeClass);
```
> Do the generated functionMethodType (in getBoundMethodHandle) always contain the boxed version of primitives? or can they contain both box and unboxed? Or do the boxed versions show up because of the asType() conversion in permuteMH? 

Nope, it won't occur in normal cases as this is not documented in the Spec (and no support from RI as such) but we can provide the users with the flexibility to choose in case that they occasionally pass over to `getBoundMethodHandle()` with the boxed types in which case we are able to handle silently by default.  If there is any updated jtreg test suite that rejects these parameters in the future, we might need to disable/remove that to match the behaviour of RI. So we don't have to intentionally optimize handle the boxed version at this point.

tajila replied:
Okay, I think it would be good to avoid the extra allocations on the native call path. 

If you change the signature of `runNativeMethod` to `long[] args` then over here you can use:
``` 
MethodHandle filterArguments​(MethodHandle target,
 int pos,
 MethodHandle... filters)
```

Where the filter is essentially the contents of `convertArgToLongValue`.  Eg.

```
if (param.type == float.class) {
     filter = Float::floatToIntBits;
} else if (param.type == double.class) {
    ...
```

For things like `((Character)argValue).charValue();` these can be handled with:

```
} else if (param.type == char.class) {
    filter = ProgrammableInvoker::c2l
} else ...


static final long c2l(char c) {
     return (long) c; //i2l bytecode
}
```

tajila replied:
with this approach we should also cache the MH for conversion so we are not generating anonClasses on the call path

ChengJin01 replied:
It makes more sense to call `filterArguments​`. Will replace the code involved to see how it goes this way.

ChengJin01 replied:
Updated with `filterArguments​` in `permuteMH()` which works good in the existing test cases

ChengJin01 commented on jcl/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableInvoker.java:
```diff
@@ -0,0 +1,461 @@
+	private void checkIfValidLayoutAndType(MethodType targetMethodType, FunctionDescriptor funcDesc) {
+		Class<?> retType = targetMethodType.returnType();
+		if (!validateArgRetTypeClass(retType) && (retType != void.class)) {
+			throw new IllegalArgumentException("The return type is neither primitive/void nor MemoryAddress");  //$NON-NLS-1$
```
It definitely works with `UnsupportedOperationException` but `IllegalArgumentException` is the expected exception of RI for any type of mismatch including the unsupported/unwanted types (e.g. a String.class).

/openj9-openjdk-jdk16/test/jdk/java/foreign/TestIllegalLink.java
```
    public void testTypeMismatch(MethodType mt, FunctionDescriptor desc, String expectedExceptionMessage) {
        try {
            ABI.downcallHandle(dummyTarget, mt, desc);
            fail("Expected IllegalArgumentException was not thrown");
        } catch (IllegalArgumentException e) {  <-------------------
            assertTrue(e.getMessage().contains(expectedExceptionMessage));
        }
    }
```
So we need to match the behaviour of RI in such case even though we could end up with different error messages.

ChengJin01 commented on jcl/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableInvoker.java:
```diff
@@ -0,0 +1,461 @@
+			 * Note: an int method (e.g. '(int,int)int') and a method function (e.g.'(boolean,boolean)boolean') share
+			 * the same C_INT layouts (e.g. 'b32[abi/kind=INT]b32[abi/kind=INT])b32[abi/kind=INT') but the underlying C
+			 * functions might be different from each other as follows:
+			 * e.g. 'int f1(int arg1, int arg2)' and 'bool f2(bool arg1, bool arg2)'.
```
Boolean is not documented in the Spec & no support from RI. To handle this case, we choose `C_INT` for the layout of `boolean` (32bits in size as specified in JVM Spec). However, the underlying C code are different as follows:
```
(int,int)int
int add2Ints(int intArg1, int intArg2) {
	int intSum = intArg1 + intArg2;
	return intSum;
}

(boolean,boolean)boolean
int add2BoolsWithOr(int boolArg1, int boolArg2) {
	int result = (boolArg1 || boolArg2);
	return result;
}
```
In which case we can't share the `prep_cif` as they point to different functions.

I already updated the comment in there as they are unclear/misleading.

tajila replied:
> Boolean is not documented in the Spec & no support from RI. 

my guess is its because Boolean is a not a C type, only exists in C++ (`bool`), and I believe this JEP is only concerned with C for the time being.

> In which case we can't share the prep_cif as they are correlated to different functions in our hashmaps.

Just so I understand, in the following case are the cifs shared

```
long func(long a1, long a2) {...}
long func2(long a1, long a2) {...}
```

what about this? (assuming 64bit - ILP64)

```
BOOLEAN func(BOOLEAN a1, BOOLEAN a2) {...} //BOOLEAN is defined as uintptr_t 
long func2(long a1, long a2) {...}
```

ChengJin01 replied:
> my guess is its because Boolean is a not a C type, only exists in C++ (bool), and I believe this JEP is only concerned with C for the time being.


Yes. That's what I assumed previously given `C_BOOL` was not implemented eventually in OpenJDK.

> Just so I understand, in the following case are the cifs shared
> long func(long a1, long a2) {...}
> long func2(long a1, long a2) {...}

Yes, the cifs are always shared as expected as they comes with the same argument  & return layouts.

> what about this? (assuming 64bit - ILP64)
> BOOLEAN func(BOOLEAN a1, BOOLEAN a2) {...} //BOOLEAN is defined as uintptr_t 
> long func2(long a1, long a2) {...}

The JEP only requires the size of Java carrier type to match the C layout.  Assuming `BOOLEAN` is `long` in size,  `(BOOLEAN,BOOLEAN)BOOLEAN` should correspond to `(long.class, long.class, long.class)` for the method type and `(C_LONG, C_LONG, C_LONG)` for the layout, which is the same as `long func2(long a1, long a2) {...}` in terms of layouts. 

To differentiate with these two functions with the same layouts, the key of the hashmap `cachedCifNativeThunkAddr` consist of the hashcode of the function descriptor (fundamentally the layouts) and the function address to uniquely identify the specified native function.

tajila commented on jcl/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableInvoker.java:
```diff
@@ -0,0 +1,472 @@
+		} else if (realArgTypeClass == MemoryAddress.class) {
+			return ((MemoryAddress)argValue).toRawLongValue();
+		}
+		return Long.valueOf(argValue.toString());
```
actually I think it just works out, since it is being truncated to the original size

ChengJin01 replied:
There should be no issue with signed numbers as it is. I will add one more test case (e.g. int) to double-check the result.

ChengJin01 replied:
Added 3 test cases for negative numbers which works good as expected.
```
PASSED: test_addTwoNegtiveBytes
PASSED: test_addTwoNegtiveInts
PASSED: test_addTwoNegtiveShorts
```

tajila commented on jcl/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableInvoker.java:
```diff
@@ -0,0 +1,602 @@
+			 * for the boolean function should be created in the native code.
+			 */
+			long functionAddrLongValue = functionAddr.address().toRawLongValue();
+			String funcDescHashPlusFuncAddr = funcDescriptor.hashCode() + "@" + functionAddrLongValue; //$NON-NLS-1$
```
Sorry, I'm still a little confused here. If the address is part of the key, doesn't that mean that no other function can have the same key.

This is my understanding:
```
int func1(int);
int func2(int);
```

funcDescHashPlusFuncAddr = "0x123"+ "@" + "[uniqueAddr]" //func1
funcDescHashPlusFuncAddr = "0x123"+ "@" + "[uniqueAddr]" //func2

So in this case `func1` and `func2` have the same signature, so they should share the same cif, however they will never have the same `addr` so sharing is not possible. Or did I miss something?

ChengJin01 replied:
> So in this case func1 and func2 have the same signature, so they should share the same cif, however they will never have the same addr so sharing is not possible.

Yes. The intent of combing the function address to the cache key is to guarantee the uniqueness of the keys for all function descriptors so as to differentiate `boolean` from `int` with the same `C_INT` layout or `char` from `short` with same `C_SHORT` layout. Otherwise, there is no need to append the function address. 

Technically I am not quite sure whether there is any potential issue in the future to share the `cif` when it comes to different method types with the same layouts. If not, I can simply remove the function address from the key to see whether it still works good as expected.

ChengJin01 replied:
I removed all the function address from the cache key and it seems everything works good as expected.
e.g. the following functions
``` 
int func1(int)
int func2(int)` 
boolean func3(boolean)
```
share the same cif given their have the same `C_INT` layouts. So there should be problem with the the adjustment for now.

We might need to keep any eye on this in case there is any update in the Spec as to how to handle `boolean`.

tajila commented on runtime/vm/FFITypeHelpers.hpp:
```diff
@@ -248,6 +248,87 @@ class FFITypeHelpers
+	 * @param layoutStringObject[in] An object containing the layout string describing the target type
+	 * @return The pointer to the ffi_type corresponding to the layout
+	 */
+	VMINLINE ffi_type*
```
can you please put an example of what the layout string looks like

ChengJin01 replied:
Example was added as suggested above.

ChengJin01 commented on jcl/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableInvoker.java:
```diff
@@ -0,0 +1,602 @@
+		for (int argIndex = 0; argIndex < nativeArgCount; argIndex++) {
+			argFilters[argIndex] = getArgumentFilter(argTypeClasses[argIndex]);
+		}
+		resultHandle = filterArguments(resultHandle, 0, argFilters).asType(nativeMethodType);
```
I need to double-check to see how it goes after removing `asType()`.

ChengJin01 replied:
All test cases passed after removing `asType()` in `permuteMH()`.

ChengJin01 commented on jcl/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableInvoker.java:
```diff
@@ -0,0 +1,602 @@
+		long returnVal = invokeNative(functionAddr.address().toRawLongValue(), cifNativeThunkAddr, args);
+		
+		/* Process the return value depending on the return type */
+		Object processedReturnVal = processReturnVal(returnVal);
```
Agreed and will update the code to see how it goes with a return filter .

ChengJin01 replied:
Added `getReturnValFilter()` in `permuteMH()` along with all filter methods to process the return value (verified with all test cases)

ChengJin01 commented on jcl/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableInvoker.java:
```diff
@@ -0,0 +1,602 @@
+	}
+	
+	/* Verify whether the specified type is primitive, MemoryAddress (for pointer) */
+	private static boolean validateArgRetTypeClass(Class<?> targetType) {
```
As I mention previously, none of boxed primitive types is supported by RI.
e.g
```
                MethodType mt = MethodType.methodType(int.class, int.class, Integer.class);
		FunctionDescriptor fd = FunctionDescriptor.of(C_INT, C_INT, C_INT);
		Symbol functionSymbol = nativeLib.lookup("add2Ints").get();
		MethodHandle mh = clinker.downcallHandle(functionSymbol, mt, fd);
		int result = (int)mh.invokeExact(112, 123);
```
will end up with java.lang.IllegalArgumentException as follows:
```
jdk16_hotspot_nightly/bin/java -Djava.library.path=/.../native-test-libs --add-modules jdk.incubator.foreign -Dforeign.restricted=permit  ffiTest
WARNING: Using incubator modules: jdk.incubator.foreign
Exception in thread "main" java.lang.IllegalArgumentException: Unsupported carrier: class java.lang.Integer
	at jdk.incubator.foreign/jdk.internal.foreign.abi.SharedUtils.checkCompatibleType(SharedUtils.java:208)
	at jdk.incubator.foreign/jdk.internal.foreign.abi.SharedUtils.checkFunctionTypes(SharedUtils.java:221)
	at jdk.incubator.foreign/jdk.internal.foreign.abi.x64.sysv.CallArranger.getBindings(CallArranger.java:91)
	at jdk.incubator.foreign/jdk.internal.foreign.abi.x64.sysv.CallArranger.arrangeDowncall(CallArranger.java:127)
	at jdk.incubator.foreign/jdk.internal.foreign.abi.x64.sysv.SysVx64Linker.downcallHandle(SysVx64Linker.java:93)
```

There should be no problem to disable boxed version as it is not explicitly allowed in the Spec and there is no support on RI. Will removed all code for boxed versions.

ChengJin01 replied:
All boxed version of primitive types were removed and verified with test cases.

tajila commented:
@gacholio Please review these changes

genie-openj9 commented:
Can one of the admins verify this patch?

gacholio commented on buildspecs/core.feature:
```diff
@@ -115,6 +115,7 @@ SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-excepti
 		<flag id="opt_newObjectHash" value="true"/>
 		<flag id="opt_newRomClassBuilder" value="true"/>
 		<flag id="opt_openjdkMethodhandle" value="false"/>
+		<flag id="opt_openjdkFfi" value="true"/>
```
Why is this a flag? The code is guarded by a JCL level check.

ChengJin01 replied:
It was previously used for `ProgrammableInvoker` and its fields set in the vmconstantpool.xml to skip the compilation in OpenJ9 JCL code (`java.base`) but `ProgrammableInvoker` was moved to different module (not `java.base`) in which case the flag might be outdated. I need to compile without the flag to see whether it is still required or not.

ChengJin01 replied:
Removed the flag as it is not needed any more.

ChengJin01 replied:
@gacholio,  I reviewed previous discussion with Pete offline on Slack.    So the reason for the flag being added in the constant pool is to indicate that all entries with `ProgrammableInvoker`  in the `vmconstantpool.xml` are not part of `java.base` and should be skipped in `initializeKnownClasses`; otherwise, it goes wrong in multiple runs with the shared class enabled because its module is incorrectly set to `java.base` in `initializeKnownClasses` at startup after the flag was removed (tried another time and it ended with the exception as follows):
```
Exception in thread "main" java.lang.IllegalAccessError: Class jdk/internal/foreign/abi/x64/sysv/CallArranger(jdk.incubator.foreign) can not access class jdk/internal/foreign/abi/ProgrammableInvoker(java.base) because module java.base does not export package jdk/internal/foreign/abi to module jdk.incubator.foreign
	at jdk.incubator.foreign/jdk.internal.foreign.abi.x64.sysv.CallArranger.arrangeDowncall(CallArranger.java:135)
	at jdk.incubator.foreign/jdk.internal.foreign.abi.x64.sysv.SysVx64Linker.downcallHandle(SysVx64Linker.java:93)
```
In this case, the flag (for both UMA & Cmake) must be added back to ensure it works good as expected.

ChengJin01 replied:
Confirmed that the exception was gone with the flag enabled there after verifying the test cases many times.

gacholio commented on runtime/vm/FFITypeHelpers.hpp:
```diff
@@ -248,6 +248,89 @@ class FFITypeHelpers
+			typeFFI = &ffi_type_sint32;
+			break;
+		case 'L':	/* either C_LONG or C_LONG_LONG */
+			typeFFI = &ffi_type_sint64;
```
What does L even mean for passing out to pure C code?

ChengJin01 replied:
We use  `L` (8 bytes) to stands for `long` (`C_LONG`) on non-Windows platforms or `C_LONG_LONG` only on Windows. I added comment there for explanation.

gacholio replied:
How does this differ from J?

ChengJin01 replied:
The code there is to parse/identify the argument type base with the 1st letter of the extract layout string of function descriptor (specified in the new feature).  e.g. The long type corresponds to `b64[abi/kind=LONG]` on Linux and `b64[abi/kind=LONG_LONG]` on Windows, in which case `L` represents the long type.

We are able to change to `J` by replacing the layout string at Java level but there is no need to to do given we have to match exactly what happens to the layout string at this point.

gacholio replied:
Is this based on the RI? This is a senseless inconsistency given that all the other characters match java (with the exception of P).

ChengJin01 replied:
This is based on the API Spec at https://docs.oracle.com/en/java/javase/16/docs/api/jdk.incubator.foreign/jdk/incubator/foreign/CLinker.html#C_LONG, https://docs.oracle.com/en/java/javase/16/docs/api/jdk.incubator.foreign/jdk/incubator/foreign/CLinker.html#C_LONG_LONG as well as https://github.com/openjdk/panama-foreign/blob/b068d7d4041bb66ddef1f0a53a35f3cebf009026/doc/panama_ffi.md.

gacholio replied:
Very annoying (though in retrospect, it was silly to choose L for objects and J for long, the opposite would have made infinitely more sense).

ChengJin01 replied:
We will need to preprocess the layout string of struct at phase 2 for easy handling in native due to complexity of struct string, in which case we are able to convert the layout string to a single letter (e.g. `LONG` or `LONG_LONG` maps to `J` for the convention in VM) if you prefer to do so.

gacholio commented on runtime/vm/FFITypeHelpers.hpp:
```diff
@@ -248,6 +248,89 @@ class FFITypeHelpers
+	 * @return The pointer to the ffi_type corresponding to the layout
+	 */
+	VMINLINE ffi_type*
+	getPrimitiveFFIType(j9object_t layoutStringObject) {
```
{ on next line.

ChengJin01 replied:
Fixed.

gacholio commented on runtime/vm/FFITypeHelpers.hpp:
```diff
@@ -248,6 +248,89 @@ class FFITypeHelpers
+														J9_STR_NULL_TERMINATE_RESULT,
+														"", 0, layoutBuffer, sizeof(layoutBuffer), NULL);
+		if (NULL == layoutStr) {
+			goto doneGetPrimitiveFFIType;
```
Labels are function-local, so this could just be called "done".

ChengJin01 replied:
Replaced with `done` as suggested.

gacholio commented on runtime/vm/FFITypeHelpers.hpp:
```diff
@@ -248,6 +248,89 @@ class FFITypeHelpers
+		 * Thus, we need to skip over '=' to obtain the first character of the type.
+		 */
+		currentLayout = layoutStr;
+		while (*currentLayout != '=') {
```
Please reverse the test.

ChengJin01 replied:
Fixed.

gacholio commented on runtime/vm/FFITypeHelpers.hpp:
```diff
@@ -248,6 +248,89 @@ class FFITypeHelpers
+		 */
+		currentLayout = layoutStr;
+		while (*currentLayout != '=') {
+			currentLayout++;
```
+= 1 is preferred.

ChengJin01 replied:
Fixed.

gacholio commented on runtime/vm/FFITypeHelpers.hpp:
```diff
@@ -248,6 +248,89 @@ class FFITypeHelpers
+		while (*currentLayout != '=') {
+			currentLayout++;
+		}
+		primitiveTypeFFI = mapPrimitiveLayoutToFFIType(*(++currentLayout));
```
Please do a +=1 on the previous line, or just use currentLayout[1] since is appears the variable is not used again.

ChengJin01 replied:
Added `currentLayout += 1;` before `mapPrimitiveLayoutToFFIType` as suggested above.

gacholio commented on runtime/vm/NativeHelpers.cpp:
```diff
@@ -149,6 +149,17 @@ convertCStringToByteArray(J9VMThread *currentThread, const char *cString)
+convertToNativeArgArray(J9VMThread *currentThread, j9object_t argArray, UDATA *javaArgs)
+{
+	UDATA argCount = (UDATA)J9INDEXABLEOBJECT_SIZE(currentThread, argArray);
+	UDATA *result = javaArgs;
```
This variable seems unnecessary.

ChengJin01 replied:
Removed this variable.

gacholio commented on runtime/vm/NativeHelpers.cpp:
```diff
@@ -149,6 +149,17 @@ convertCStringToByteArray(J9VMThread *currentThread, const char *cString)
+UDATA *
+convertToNativeArgArray(J9VMThread *currentThread, j9object_t argArray, UDATA *javaArgs)
+{
+	UDATA argCount = (UDATA)J9INDEXABLEOBJECT_SIZE(currentThread, argArray);
```
This could be inside the NULL test.

ChengJin01 replied:
The code there has been updated to remove the NULL check.

gacholio commented on runtime/vm/NativeHelpers.cpp:
```diff
@@ -149,6 +149,17 @@ convertCStringToByteArray(J9VMThread *currentThread, const char *cString)
+	UDATA argCount = (UDATA)J9INDEXABLEOBJECT_SIZE(currentThread, argArray);
+	UDATA *result = javaArgs;
+	if (NULL != result) {
+		VM_ArrayCopyHelpers::memcpyFromArray(currentThread, argArray, (UDATA)3, 0, argCount, (void*)result);
```
UDATA cast is unnecessary.

ChengJin01 replied:
Removed.

gacholio commented on runtime/vm/NativeHelpers.cpp:
```diff
@@ -149,6 +149,17 @@ convertCStringToByteArray(J9VMThread *currentThread, const char *cString)
+{
+	UDATA argCount = (UDATA)J9INDEXABLEOBJECT_SIZE(currentThread, argArray);
+	UDATA *result = javaArgs;
+	if (NULL != result) {
```
The doc says this can't be NULL.

ChengJin01 replied:
The code there has been updated to remove the NULL check.

gacholio commented on runtime/vm/NativeHelpers.cpp:
```diff
@@ -149,6 +149,17 @@ convertCStringToByteArray(J9VMThread *currentThread, const char *cString)
+	UDATA argCount = (UDATA)J9INDEXABLEOBJECT_SIZE(currentThread, argArray);
+	UDATA *result = javaArgs;
+	if (NULL != result) {
+		VM_ArrayCopyHelpers::memcpyFromArray(currentThread, argArray, (UDATA)3, 0, argCount, (void*)result);
```
Some description of the magic "3" would be nice.

ChengJin01 replied:
Comment was added for `3` (3 means 8 bytes for each element of the array to be copied).

gacholio commented on runtime/oti/j9nonbuilder.h:
```diff
@@ -5004,6 +5004,10 @@ typedef struct J9VMThread {
 #endif /* OMR_GC_CONCURRENT_SCAVENGER */
 	UDATA safePointCount;
+#if JAVA_SPEC_VERSION >= 16
+	UDATA *javaArgs;
```
These aren't great names. I thought this was a thread local copy of the arguments to the java executeable.

ChengJin01 replied:
Renamed to `ffiArgs` and `ffiArgCount` which are specific to the parameters passed over to `ffi_call`.

gacholio commented on runtime/vm/NativeHelpers.cpp:
```diff
@@ -149,6 +149,17 @@ convertCStringToByteArray(J9VMThread *currentThread, const char *cString)
+convertToNativeArgArray(J9VMThread *currentThread, j9object_t argArray, UDATA *ffiArgs)
+{
+	UDATA argCount = (UDATA)J9INDEXABLEOBJECT_SIZE(currentThread, argArray);
+	/* 3 means each element of the array to be copied is 8 bytes (64bits) in size
```
So does this mean the type is not actually `UDATA*`? Or does the value need to change between 32 and 64 bit?

ChengJin01 replied:
We already covert all parameters to `long` at first in `ProgrammableInvoker` do which means it must be 64 bit in there.

gacholio replied:
Then the type should really be `jlong`.

gacholio replied:
You're getting lucky because I bet you've only run this on a little endian platform. The array is being accessed as UDATA which will certainly not work in most cases.

ChengJin01 replied:
> Then the type should really be jlong

I am not quite following this (why it should be `jlong` which is a JNI type). 

> You're getting lucky because I bet you've only run this on a little endian platform. The array is being accessed as UDATA which will certainly not work in most cases.

Actually we have code partially set up on `AIX/PPC` which works good (except the Valist support) without any error.
We already converted the parameter to specific types required in ffi_call as follows:
```
inlProgrammableInvokerInvokeNative
#if !defined(J9VM_ENV_LITTLE_ENDIAN)
				/* Note: A float number is converted to int by Float.floatToIntBits() in ProgrammableInvoker */
				if ((J9NtcInt == argType) || (J9NtcBoolean == argType) || (J9NtcFloat == argType)) {
					values[i] = (void *)((UDATA)values[i] + extraBytesOfInt);
				} else if ((J9NtcShort == argType) || (J9NtcChar == argType)) {
					values[i] = (void *)((UDATA)values[i] + extraBytesOfShortAndChar);
				} else if (J9NtcByte == argType) {
					values[i] = (void *)((UDATA)values[i] + extraBytesOfByte);
				}
```

gacholio replied:
I misspoke - it won't work on 32 bit where UDATA is only 4 bytes.

ChengJin01 replied:
I don't think we need to support 32 bit. 
Am I correct at this point, @tajila?

gacholio replied:
Even if we decide that, you should use the correct type U_64 instead of UDATA.

ChengJin01 replied:
Agreed and replaced with `U_64` in such case.

tajila replied:
There are currently no plans to support 32bit for JDK11+ at the moment. However, there is a small possibility that we may need to support z/OS-31bit.

ChengJin01 replied:
There is no code changes for now to address this case. So might need to add a little code when it comes up.

gacholio commented on runtime/vm/jvmfree.c:
```diff
@@ -267,6 +267,13 @@ deallocateVMThread(J9VMThread * vmThread, UDATA decrementZombieCount, UDATA send
 #endif
 
+#if JAVA_SPEC_VERSION >= 16
+	if (NULL != vmThread->ffiArgs) {
```
Free of NULL is allowed, so you could remove this check (I believe the expected case would be that it's not NULL).

ChengJin01 replied:
Removed the NULL check as suggested above.

gacholio commented on runtime/vm/vm_internal.h:
```diff
@@ -431,6 +431,18 @@ convertByteArrayToCString(J9VMThread *currentThread, j9object_t byteArray);
+ * @returns the newly-allocated memory, or NULL on failure (no exception is set pending)
+ */
+UDATA *
+convertToNativeArgArray(J9VMThread *currentThread, j9object_t argArray, UDATA *javaArgs);
```
I believe these should also be U64.

gacholio replied:
Ditto for the function declaration.

ChengJin01 replied:
Replaced with `U_64`.

keithc-ca replied:
This is still `UDATA`, not `U_64`. Did I miss some discussion that concluded differently?

gacholio commented:
jenkins test sanity zlinux jdk8

gacholio commented:
jenkins test sanity,extended aix jdk16

gacholio commented:
jenkins compile win jdk8

gacholio commented:
jenkins test sanity,extended win jdk16

gacholio commented:
Java compile failures in JDK8.

ChengJin01 commented:
> Java compile failures in JDK8.

This changes depend on the OpenJDKMH changes which means it requires something like `jenkins test sanity winojdk292,xlinuxojdk292 jdk16 depends ibmruntimes/openj9-openjdk-jdk16#29`

gacholio commented:
The java code must compile without requiring that depends, or this cant be merged.

ChengJin01 commented:
> 
> 
> The java code must compile without requiring that depends, or this cant be merged.

The required changes are on OpenJDK side at https://github.com/ibmruntimes/openj9-openjdk-jdk16/pull/46

tajila commented:
jenkins test sanity,extended win jdk16

tajila commented:
jenkins test sanity,extended aix jdk16

tajila commented:
jenkins test sanity zlinux jdk8

gacholio commented:
The s390 failure appears to be infra-related.

keithc-ca commented on test/functional/Java16andUp/src/org/openj9/test/jep389/downcall/InvalidDownCallTests.java:
```diff
@@ -0,0 +1,387 @@
+ */
+@Test(groups = { "level.sanity" })
+public class InvalidDownCallTests {
+	private static boolean isWinOS = System.getProperty("os.name").toLowerCase().contains("win") ? true : false;
```
Please remove `? true : false` (it's simply redundant).

ChengJin01 replied:
Agreed and fixed as suggested above.

keithc-ca commented on test/functional/Java16andUp/src/org/openj9/test/jep389/downcall/PrimitiveTypeTests.java:
```diff
@@ -0,0 +1,588 @@
+ */
+@Test(groups = { "level.sanity" })
+public class PrimitiveTypeTests {
+	private static boolean isWinOS = System.getProperty("os.name").toLowerCase().contains("win") ? true : false;
```
Please remove `? true : false`.

ChengJin01 replied:
Fixed as suggested above.

keithc-ca commented on test/functional/Java16andUp/src/org/openj9/test/jep389/downcall/PrimitiveTypeTests.java:
```diff
@@ -0,0 +1,588 @@
+		FunctionDescriptor fd = FunctionDescriptor.of(C_INT, C_INT, C_VA_LIST);
+		NativeScope nativeScope = NativeScope.unboundedScope();
+		VaList vaList = CLinker.VaList.make(vaListBuilder -> vaListBuilder.vargFromInt(C_INT, 700)
+															.vargFromInt(C_INT, 800)
+															.vargFromInt(C_INT, 900)
+															.vargFromInt(C_INT, 1000),
+															nativeScope);
```
Excessive indentation.

ChengJin01 replied:
Fixed at all places.

keithc-ca replied:
I think the line-breaks made for easier reading. I'd like to seem them return, but don't indent more than one tab extra compared with the start of the declaration:
```
		VaList vaList = CLinker.VaList.make(vaListBuilder -> vaListBuilder
			.vargFromInt(C_INT, 700)
			.vargFromInt(C_INT, 800)
			.vargFromInt(C_INT, 900)
			.vargFromInt(C_INT, 1000), nativeScope);
```

ChengJin01 replied:
Fixed as suggested above.

ChengJin01 commented on buildspecs/core.feature:
```diff
@@ -1,6 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <!--
-Copyright (c) 2006, 2020 IBM Corp. and others
+Copyright (c) 2006, 2021 IBM Corp. and others
```
Reverted.

ChengJin01 commented on jcl/.classpath:
```diff
@@ -38,6 +38,7 @@
 	<classpathentry excluding="**/module-info.java" kind="src" path="src/openj9.jvm/share/classes"/>
 	<classpathentry excluding="**/module-info.java" kind="src" path="src/openj9.sharedclasses/share/classes"/>
 	<classpathentry excluding="**/module-info.java" kind="src" path="src/openj9.traceformat/share/classes"/>
+	<classpathentry excluding="**/module-info.java" kind="src" path="src/jdk.incubator.foreign/share/classes"/>
```
Fixed.

ChengJin01 commented on runtime/tests/clinkerffi/downcall.c:
```diff
@@ -0,0 +1,318 @@
+ * @param intArg2 the 2nd integer to add
+ * @return the sum of the two integers
+ */
+int add2Ints(int intArg1, int intArg2) {
```
Fixed at all places.

keithc-ca replied:
It appears that `add2Ints` is the only function you didn't fix.

ChengJin01 replied:
Fixing the overlooked one.

ChengJin01 commented on runtime/tests/clinkerffi/module.xml:
```diff
@@ -0,0 +1,71 @@
+   [1] https://www.gnu.org/software/classpath/license.html
+   [2] http://openjdk.java.net/legal/assembly-exception.html
+
+   SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
```
Fixed.

ChengJin01 commented on runtime/vm/BytecodeInterpreter.hpp:
```diff
@@ -4669,6 +4669,182 @@ class INTERPRETER_CLASS
+
+		j9object_t argValues = *(j9object_t *)_sp;					// argValues
+		ffi_cif *cif = (ffi_cif *)(UDATA)*(I_64 *)(_sp + 1);		// calloutThunk
+		void *function = (void *)(UDATA)*(I_64 *)(_sp + 3);			// functionAddr
```
Fixed.

ChengJin01 commented on jcl/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableInvoker.java:
```diff
@@ -0,0 +1,565 @@
+			argLayoutStrs[argIndex] = argLayout.toString();
+		}
+		
+		synchronized(privateClassLock) {
```
Fixed.

ChengJin01 commented on jcl/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableInvoker.java:
```diff
@@ -0,0 +1,565 @@
+	/* The method is ultimately invoked by Clinker on the specific platforms to generate the requested
+	 * method handle to the underlying C function.
+	 */
+	@SuppressWarnings("javadoc")
```
Declarations are added in there.
Please note this is not a public API for users (only for internal)

keithc-ca replied:
The documentation should be in a doc-comment (beginning with `/**`).

ChengJin01 replied:
Fixed.

pshipton commented:
Opened issue for infra problem and disabled the machine. https://github.com/eclipse-openj9/openj9/issues/12713

keithc-ca commented on jcl/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableInvoker.java:
```diff
@@ -0,0 +1,570 @@
+	final static Lookup lookup = MethodHandles.lookup();
+	
+	/* The prep_cif and the corresponding argument layouts are cached & shared in multiple downcalls/threads */
+	private static final HashMap<Integer, Long> cachedCifNativeThunkAddr = new HashMap<Integer, Long>();
+	private static final HashMap<List<MemoryLayout>, Long> cachedArgLayouts = new HashMap<List<MemoryLayout>, Long>();
```
The type arguments can be inferred, you can write just `new HashMap<>()`.

ChengJin01 replied:
Nope, it will mess up the key & value inside there.

keithc-ca replied:
I don't understand why you say that. This compiles just fine.
```
	private static final HashMap<Integer, Long> cachedCifNativeThunkAddr = new HashMap<>();
	private static final HashMap<List<MemoryLayout>, Long> cachedArgLayouts = new HashMap<>();
```

ChengJin01 replied:
This has nothing to do with the compilation. I mean the test results will be messed up with the changes above (used to tried with this but ended up screwed up everything in tests).  I will double-check this to see whether it works as the key & value are already modified.

keithc-ca replied:
It should change nothing, except perhaps avoid a warning from `javac`.

ChengJin01 replied:
Replaced with `HashMap<>()` and verified all test cases.

keithc-ca commented on jcl/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableInvoker.java:
```diff
@@ -0,0 +1,570 @@
+	private static final HashMap<List<MemoryLayout>, Long> cachedArgLayouts = new HashMap<List<MemoryLayout>, Long>();
+
+	/* Argument filters that convert the primitive types or MemoryAddress to long */
+	private static MethodHandle booleanToLongArgFilter = null;
+	private static MethodHandle charToLongArgFilter = null;
+	private static MethodHandle byteToLongArgFilter = null;
+	private static MethodHandle shortToLongArgFilter = null;
+	private static MethodHandle intToLongArgFilter = null;
+	private static MethodHandle floatToLongArgFilter = null;
+	private static MethodHandle doubleToLongArgFilter = null;
+	private static MethodHandle memAddrToLongArgFilter = null;
+	
+	/* Return value filters that convert the Long object to the primitive types or MemoryAddress */
+	private static MethodHandle longObjToVoidRetFilter = null;
+	private static MethodHandle longObjToBooleanRetFilter = null;
+	private static MethodHandle longObjToCharRetFilter = null;
+	private static MethodHandle longObjToByteRetFilter = null;
+	private static MethodHandle longObjToShortRetFilter = null;
+	private static MethodHandle longObjToIntRetFilter = null;
+	private static MethodHandle longObjToLongRetFilter = null;
+	private static MethodHandle longObjToFloatRetFilter = null;
+	private static MethodHandle longObjToDoubleRetFilter = null;
+	private static MethodHandle longObjToMemAddrRetFilter = null;
```
Initializing all these fields is at best redundant. Can they be made `final` (they all appear to be initialized in `<clinit>`).

ChengJin01 replied:
Fixed as suggested above.

keithc-ca commented on jcl/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableInvoker.java:
```diff
@@ -0,0 +1,570 @@
+		return filterMH;
+	}
+	
+	/* The return value filter that converts the returned long value from the C function to the specified return type at Java level */
```
This line is too long. It's not possible to read it on the web without horizontal scrolling.

ChengJin01 replied:
Fixed as suggested above.

keithc-ca commented on jcl/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableInvoker.java:
```diff
@@ -0,0 +1,570 @@
+		Class<?> retType = targetMethodType.returnType();
+		if (!validateArgRetTypeClass(retType) && (retType != void.class)) {
+			throw new IllegalArgumentException("The return type must be primitive/void or MemoryAddress" //$NON-NLS-1$
+												+ ": retType = " + retType);  //$NON-NLS-1$
```
Excessive indentation (here and lines 406-407 and elsewhere).

ChengJin01 replied:
Fixed at all places as suggested above.

keithc-ca commented on jcl/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableInvoker.java:
```diff
@@ -0,0 +1,570 @@
+		}
+		
+		Optional<MemoryLayout> returnLayout = funcDesc.returnLayout();
+		realReturnLayout = returnLayout.isPresent() ? returnLayout.get() : null; // set to null for void
```
This can use `returnLayout.orElse(null)`.

ChengJin01 replied:
Updated as suggested above and need to compile to verify the test cases.

ChengJin01 replied:
All test cases passed with the changes

keithc-ca commented on jcl/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableInvoker.java:
```diff
@@ -0,0 +1,570 @@
+	
+	/* Check the size of the specified primitive layout to determine whether it matches the specified type */
+	private static void validateValueLayoutSize(MemoryLayout TypeLayout, Class<?> targetType) {
+		String layoutSize = TypeLayout.bitSize() + "_bits"; //$NON-NLS-1$
```
This seems a rather expensive way to handle certain values (8, 16, 32, 64) of a `long`.

ChengJin01 replied:
Agreed and replaced with directly integer to check the bit size.

keithc-ca commented on jcl/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableInvoker.java:
```diff
@@ -0,0 +1,570 @@
+	private static void validateValueLayoutKind(MemoryLayout targetLayout, Class<?> targetType) {
+		boolean mismatchType = false;
+		
+		if(!targetLayout.toString().contains("[abi/kind=")) { //$NON-NLS-1$
```
formatting: missing space after `if`

ChengJin01 replied:
Fixed.

keithc-ca commented on jcl/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableInvoker.java:
```diff
@@ -0,0 +1,570 @@
+		 * e.g. b32[abi/kind=INT]
+		 */
+		TypeKind kind = (TypeKind)targetLayout.attribute(TypeKind.ATTR_NAME).orElse(null);
+		switch (kind) {
```
I believe a `NullPointerException` will result if `kind == null`.

ChengJin01 replied:
Not really because `abi/kind` must be one of the following layout name in the switch according to the Spec at https://github.com/openjdk/panama-foreign/blob/b068d7d4041bb66ddef1f0a53a35f3cebf009026/doc/panama_ffi.md.
```
C layout 	Java carrier
C_BOOL 	byte
C_CHAR 	byte
C_SHORT 	short
C_INT 	int
C_LONG 	long
C_LONGLONG 	long
C_FLOAT 	float
C_DOUBLE 	double
C_POINTER 	MemoryAddress
GroupLayout 	MemorySegment
C_VALIST 	CLinker.VaList
```

keithc-ca replied:
Then the use of `orElse(null)` is inappropriate. It should use `orElseThrow()`, perhaps like this:
```
TypeKind kind = (TypeKind)targetLayout.attribute(TypeKind.ATTR_NAME).orElseThrow(() -> new IllegalArgumentException());
```

ChengJin01 replied:
Agreed and will need to verify the test cases with the changes.

ChengJin01 replied:
All test cases passed with the changes.

keithc-ca commented on runtime/vm/BytecodeInterpreter.hpp:
```diff
@@ -4669,6 +4669,182 @@ class INTERPRETER_CLASS
+		/* Make sure we can fit a double in each sValues_raw[] slot but assuring we end up
+		 * with an int that is a  multiple of sizeof(ffi_raw)
+		 */
+		const U_8 valRawWorstCaseMulFactor = ((sizeof(double) - 1U)/sizeof(ffi_raw)) + 1U;
```
missing spaces surrounding `/`

ChengJin01 replied:
Fixed.

keithc-ca commented on jcl/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableInvoker.java:
```diff
@@ -0,0 +1,571 @@
+		 * e.g. b32[abi/kind=INT]
+		 */
+		TypeKind kind = (TypeKind)targetLayout.attribute(TypeKind.ATTR_NAME)
+				.orElseThrow(() -> new IllegalArgumentException("The layout's ABI Class is empty: layout = " + targetLayout)); //$NON-NLS-1$
```
I don't think there's much value in appending the empty optional in the message; I suggest the message can be simplified to `"The layout's ABI class is empty"`.

ChengJin01 replied:
Updated the message.

ChengJin01 commented on test/functional/Java16andUp/build.xml:
```diff
@@ -76,7 +76,7 @@
 	</target>
 
-	<target name="compile" depends="init,getDependentLibs" description="Using java ${JDK_VERSION} to compile the source" >
+	<target name="compile" depends="init" description="Using java ${JDK_VERSION} to compile the source" >
```
I am not sure whether it is still required. @llxia,  any insight on this?

llxia replied:
`getDependentLibs` is needed for downloading 3rd party libs in line https://github.com/eclipse-openj9/openj9/blob/234e63dae833dcbd739ad9fac4e611b3eb85a486/test/functional/Java16andUp/build.xml#L41

ChengJin01 replied:
Added it back as requested above.

ChengJin01 reviewed (commented):
I updated all the code & test cases against comments from @keithc-ca (didn't show up in notes previously).

ChengJin01 commented on jcl/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableInvoker.java:
```diff
@@ -0,0 +1,571 @@
+	
+	/* Check the size of the specified primitive layout to determine whether it matches the specified type */
+	private static void validateValueLayoutSize(MemoryLayout TypeLayout, Class<?> targetType) {
+		int layoutSize = (int)TypeLayout.bitSize();
```
Added comments there to clarify this situation for primitives.

ChengJin01 commented on buildenv/.project:
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<projectDescription>
+	<name>buildenv</name>
+	<comment></comment>
```
Removed.

ChengJin01 commented on jcl/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableInvoker.java:
```diff
@@ -0,0 +1,573 @@
+		} catch (NoSuchMethodException e) {
+			throw new RuntimeException("Failed to find the requested method: " + e.getMessage()); //$NON-NLS-1$
+		} catch (IllegalAccessException e) {
+			throw new RuntimeException("Illegal access to the requested method: " + e.getMessage()); //$NON-NLS-1$
```
Updated as suggested above.

ChengJin01 commented on jcl/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableInvoker.java:
```diff
@@ -0,0 +1,573 @@
+
+	/* Intended for longObjToBooleanRetFilter that converts the Long object to boolean */
+	private static final boolean longObjToBooleanRet(Object retValue) {
+		boolean resultValue = (((Long)retValue).intValue() == 1) ? true : false;
```
Updated as suggested above.

ChengJin01 commented on jcl/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableInvoker.java:
```diff
@@ -0,0 +1,573 @@
+			 */
+			int funcDescHash = funcDescriptor.hashCode();
+			if (cachedCifNativeThunkAddr.containsKey(funcDescHash)) {
+				cifNativeThunkAddr = cachedCifNativeThunkAddr.get(funcDescHash).longValue();
```
Updated as suggested above.

llxia commented on test/functional/Java16andUp/playlist.xml:
```diff
@@ -116,6 +116,32 @@
+	</test>
+	
+	<test>
+		<testCaseName>Jep389Tests_testClinkerFfi</testCaseName>
```
Is this test specific to openj9/ibm? If so, please add `<impl>`. Thanks

ChengJin01 replied:
Nope. It is intended for OpenJ9/OpenJDK at least for now.

llxia commented on test/functional/Java16andUp/testClinkerFfi.xml:
```diff
@@ -0,0 +1,38 @@
+			<class name="org.openj9.test.jep389.downcall.MultiThreadingTests4"/>
+			<class name="org.openj9.test.jep389.downcall.InvalidDownCallTests"/>
+		</classes>
+	</test>
```
Can we add this change in https://github.com/eclipse-openj9/openj9/blob/be92fa66592ddaeee8ce444f0c7b11ac95a893a1/test/functional/Java16andUp/testng.xml?

ChengJin01 replied:
Nope, this is for the new feature (JEP389) which has nothing to with JEP397.

keithc-ca commented on runtime/vm/FFITypeHelpers.hpp:
```diff
@@ -248,6 +248,91 @@ class FFITypeHelpers
+		 * Thus, we need to skip over '=' to obtain the first character of the type.
+		 */
+		currentLayout = layoutStr;
+		while ('=' != *currentLayout) {
```
Who validate the string (e.g. that is contains `'='`)?

ChengJin01 replied:
This is a specified converted layout strings from Java code (converting directly from layout to string) e.g.  `C_INT` is `C_INT = b32[abi/kind=INT]` which is already set up in OpenJDK. That being said, this is not a generic string to be validated; otherwise, any error in the layout string will be captured in Java.

keithc-ca replied:
Actually, I now see the check in `validateValueLayoutKind()`:
```
	if (!targetLayout.toString().contains("[abi/kind=")) { //$NON-NLS-1$
		throw new IllegalArgumentException("The layout's ABI Class is undefined: layout = " + targetLayout); //$NON-NLS-1$
	}
```

llxia commented on test/functional/Java16andUp/playlist.xml:
```diff
@@ -116,6 +116,32 @@
 			-excludegroups $(DEFAULT_EXCLUDE); \
 			$(TEST_STATUS)
 		</command>
+		<platformRequirements>bits.64,^arch.ppc,^arch.390,^arch.riscv,^os.zos,^os.sunos,^arch.arm</platformRequirements>
```
just to double check, is it intended to add `platformRequirements` for `Jep397Tests_testSubClassOfSealedSuperFromDifferentPackageInSameNamedModule`?

ChengJin01 replied:
The `platformRequirements` is temporarily set up for JEP389 (which is irrelevant to JEP397) to disable a couple of platforms we have not yet implemented but will do in the future as scheduled.

ChengJin01 commented:
> This is still UDATA, not U_64. Did I miss some discussion that concluded differently?

Fixed as ignored previously.

llxia reviewed (approved):
The test change looks good to me.

ChengJin01 commented:
@tajila, please merge the approved changes along with the change in OpenJDK at https://github.com/ibmruntimes/openj9-openjdk-jdk16/pull/46 so we can get started to prepare for the code at phase 2 & 3 on the top of the changes there.

keithc-ca commented:
ibmruntimes/openj9-openjdk-jdk16#46 was reverted: we'll need a new PR to replace it.

ChengJin01 commented:
> 
> 
> [ibmruntimes/openj9-openjdk-jdk16#46](https://github.com/ibmruntimes/openj9-openjdk-jdk16/pull/46) was reverted: we'll need a new PR to replace it.

A new PR in OpenJDK16 has been created at https://github.com/ibmruntimes/openj9-openjdk-jdk16/pull/53.

tajila commented:
Jenkins test sanity plinux jdk16 depends ibmruntimes/openj9-openjdk-jdk16#53

tajila commented:
Jenkins test sanity plinux jdknext depends https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/314

tajila commented:
Jenkins test sanity win jdknext depends ibmruntimes/openj9-openjdk-jdk#314

tajila commented:
Jenkins test sanity win jdknext depends ibmruntimes/openj9-openjdk-jdk#314

tajila commented:
Jenkins test sanity plinux jdknext depends ibmruntimes/openj9-openjdk-jdk16#53

tajila commented:
Jenkins test sanity plinux jdk16 depends ibmruntimes/openj9-openjdk-jdk16#53

ChengJin01 commented:
The failure at https://openj9-jenkins.osuosl.org/job/Test_openjdknext_j9_sanity.functional_x86-64_windows_Personal_testList_0/1/consoleText indicates the build didn't pick up the correct platform suffixed with `_ojdk292` changes on OpenJDKMH.
e.g. it should be `x86-64_linux_ojdk292` on Linux and `x86-64_windows_ojdk292` on Windows, in which case it ended up with the exception as follows due to the missing OpenJDKMH changes:
```
Caused by: java.lang.ClassCastException: java.lang.invoke.DirectHandle incompatible with java.lang.invoke.BoundMethodHandle
...
```

gacholio reviewed (approved):
Someone else can merge this along with the other required PRs.

tajila commented:
Looks like we will need to wait until ojdk_mh is enabled before merging this. We have a tentative date for June 7 for enablement.

tajila commented:
@ChengJin01 Please rebase these changes. OJDK MH is enabled on jdk16, I will re-run the PR builds once you've rebased.

ChengJin01 commented:
@tajila, the changes have been rebased (conflict resolved). Please go ahead to run jenkins tests to verify the changes.

tajila commented:
jenkins test sanity xlinuxojdk292 jdk16 depends ibmruntimes/openj9-openjdk-jdk16#53

tajila commented:
jenkins test sanity winojdk292 jdk16 depends ibmruntimes/openj9-openjdk-jdk16#53

tajila commented:
Jenkins test sanity win jdknext

tajila commented:
Jenkins test sanity plinux jdk16

ChengJin01 commented:
The compilation failures on Windows at https://openj9-jenkins.osuosl.org/job/Build_JDKnext_x86-64_windows_Personal/16/console indicate it didn't pick up the changes in OpenJDK extension at https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/314
```
15:14:00  === Output from failing command(s) repeated here ===
15:14:00  * For target jdk_modules_jdk.incubator.foreign__the.jdk.incubator.foreign_batch:
15:14:00  f:\users\jenkins\workspace\build_jdknext_x86-64_windows_personal\src\jdk.incubator.foreign\share\classes\jdk\internal\foreign\abi\aarch64\CallArranger.java:132: error: incompatible types: ABIDescriptor cannot be converted to Addressable
15:14:00          MethodHandle handle = new ProgrammableInvoker(C, addr, bindings.callingSequence).getBoundMethodHandle();
15:14:00                                                        ^
15:14:00  f:\users\jenkins\workspace\build_jdknext_x86-64_windows_personal\src\jdk.incubator.foreign\share\classes\jdk\internal\foreign\abi\x64\sysv\CallArranger.java:129: error: incompatible types: ABIDescriptor cannot be converted to Addressable
15:14:00          MethodHandle handle = new ProgrammableInvoker(CSysV, addr, bindings.callingSequence).getBoundMethodHandle();
15:14:00                                                        ^
15:14:00  f:\users\jenkins\workspace\build_jdknext_x86-64_windows_personal\src\jdk.incubator.foreign\share\classes\jdk\internal\foreign\abi\x64\windows\CallArranger.java:130: error: incompatible types: ABIDescriptor cannot be converted to Addressable
15:14:00          MethodHandle handle = new ProgrammableInvoker(CWindows, addr, bindings.callingSequence).getBoundMethodHandle();
```

pshipton commented:
Jenkins test sanity zlinux jdk16 depends ibmruntimes/openj9-openjdk-jdk16#53

pshipton commented:
Jenkins test sanity win jdknext depends ibmruntimes/openj9-openjdk-jdk#314

pshipton commented:
I cancelled those PR builds, I see Toby already run them.

pshipton commented:
Actually, this one is still desirable, without the ojdk292

Jenkins test sanity zlinux jdk16 depends ibmruntimes/openj9-openjdk-jdk16#53

pshipton commented:
Oops, I meant

Jenkins test sanity zlinux jdknext depends ibmruntimes/openj9-openjdk-jdk#314

tajila commented:
@ChengJin01 Can you rebase this?

ChengJin01 commented:
Rebased with all conflicts resolved in the changes.

tajila commented:
Jenkins test sanity win jdk8

tajila commented:
Jenkins test sanity.openjdk plinux jdk17

tajila commented:
Jenkins test sanity xlinux jdknext

tajila commented:
Jenkins test sanity,extended zlinux jdk17

tajila commented:
jenkins test sanity,extended xlinux jdknext depends ibmruntimes/openj9-openjdk-jdk#314

tajila commented:
Jenkins test sanity win jdk8

ChengJin01 commented:
Remove `import jdk.incubator.foreign.LibraryLookup` as it is not required in our code in JDK16 and is totally removed in JDK7 & beyond.

ChengJin01 commented:
It turns out that there were a bunch of changes recently merged to `jdk.incubator.foreign` in JDK17 & beyond which affects our downcall implementation, including:
1) `Aarch64` was split into `Linux/Aarch64` and `macOS/Aarch64`
2) `Addressable symbol` (used in our code for the native function address) was removed from the parameter list of the downcall APIs to support Hotspot.

So, we need to update the changes at ibmruntimes/openj9-openjdk-jdk#314 to get `Addressable symbol` back for our own implementation and deal with other changes accordingly. (the same changes will be created later in JDK17)

ChengJin01 commented:
The updated code is to accommodate the existing downcall implementation in JDK16 and the new API changes to support the following CLinker APIs in JDK17 & beyond:
**(1) MethodHandle downcallHandle(_Addressable symbol, SegmentAllocator allocator_, MethodType type,FunctionDescriptor function)**
e.g.
MethodHandle mh = clinker.downcallHandle(addr, allocator, mt, fd);
mh.invokeExact(argments);
Note: addr & allocator are passed in to downcallHandle() before invocation where allocator is only passed in when the return type is MemorySegment (intended for struct).

**(2) MethodHandle downcallHandle(_Addressable symbol,_ MethodType type, FunctionDescriptor function)**
e.g.
MethodHandle mh = clinker.downcallHandle(addr, mt, fd);
mh.invokeExact(argments);
Note: addr is passed to downcallHandle() before invocation (no allocator specified).

**(3) MethodHandle downcallHandle(MethodType type, FunctionDescriptor function)**
e.g.
MethodHandle mh = clinker.downcallHandle(mt, fd); 
mh.invokeExact(addr, allocator, argments);
Note: addr & allocator are passed in the invocation where allocator is only passed in when the return type is MemorySegment (intended for struct).

To fit with all these changes in `ProgrammableInvoker` in both JDK16 and JDK17 & beyond, the existing test suite in `Java16andUp` has been moved a new folder called `src_160` in in `Java16andUp` to ensure the test cases there only work for Java 16 while a copy of the test suites with the API changes has been created in `Java17andUp`.

Will need to launch personal builds to ensure all these changes work good in Java 16, 17 and beyond as expected.

ChengJin01 commented:
All updated changes (including code & test suites) were verified locally and in personal builds.

@tajila, please take a look at the updated methods in `ProgrammableInvoker` (no changes in native) which support new parameters (`Addressable` & `SegmentAllocator`) as follows:
```
1) ProgrammableInvoker(Addressable downcallAddr, MethodType functionMethodType, FunctionDescriptor functionDescriptor)
2) public static MethodHandle getBoundMethodHandle(Addressable downcallAddr, MethodType functionMethodType, FunctionDescriptor funcDesc)
3) private MethodHandle permuteMH(...)
4) Object runNativeMethod(Addressable downcallAddr, SegmentAllocator segmtAllocator, long[] args)
```

There is no fundamental change in test cases except the following adjustments for Java 16, 17 & beyond:
[1] `/test/functional/Java16andUp`
1) moved `org/openj9/test/jep389` to `src_160` as the related APIs are only supported in JDK16
2) moved `org/openj9/test/foreignMemoryAccess` to `src_160` as `jdk.internal.foreign.MemoryScope` & related APIs were totally removed in Java 16
[2] `/test/functional/Java17andUp` (newly created)
copied the test cases in  `Java16andUp/src_160/org/openj9/test/jep389` to `Java17andUp` with the following changes:
1) replaced the code of `jdk.incubator.foreign.LibraryLookup` (only in JDK16 & removed in JDK17) with `jdk.incubator.foreign.SymbolLookup` (introduced in JDK17)
2) replaced the code of `jdk.incubator.foreign.NativeScope` (only in JDK16 & removed in JDK17) with `jdk.incubator.foreign.ResourceScope` (introduced in JDK17)

@llxia, please double-check the test scripts in `Java16andUp` and `Java17andUp` to ensure it fits with the existing test framework.

I will need to bring all these changes to phase 2 &3 (struct support) and 4 (other platforms) if no more concern.

tajila commented:
jenkins test sanity,extended xlinux jdknext depends ibmruntimes/openj9-openjdk-jdk#314
